examples/vision/image_classification_efficientnet_fine_tuning.py

```diff
@@ -0,0 +1,495 @@
+variants of the model, hence the input resolution for B0 and B1 are chosen as 224 and
+240.
+
+- Depth and width: The building blocks of EfficientNet demands channel size to be
+multiples of 8.
+
+- Resource limit: Memory limitation may bottleneck resolution when depth
+and width can still increase. In such a situation, increasing depth and/or
+width but keep resolution can still improve performance.
+
+As a result, the depth, width and resolution of each variant of the EfficientNet models
+are hand-picked and proven to produce good results, though they may be significantly
+off from the compound scaling formula.
+Therefore, the keras implementation (detailed below) only provide these 8 models, B0 to B7,
+instead of allowing arbitray choice of width / depth / resolution parameters.
+
+## Keras implementation of EfficientNet
+
+An implementation of EfficientNet B0 to B7 has been shipped with tf.keras since TF2.3. To
+use EfficientNetB0 for classifying 1000 classes of images from imagenet, run
+```
+from tensorflow.keras.applications import EfficientNetB0
+model = EfficientNetB0(weights='imagenet')
+```
+
+This model takes input images of shape (224, 224, 3), and the input data should range
+[0, 255]. Normalization is included as part of the model.
+
+Because training EfficientNet on ImageNet takes a tremendous amount of resources and
+several techniques that are not a part of the model architecture itself. Hence the Keras
+implementation by default loads pre-trained weights obtained via training with
+[AutoAugment](https://arxiv.org/abs/1805.09501).
+
+For B0 to B7 base models, the input shapes are different. Here is a list of input shape
+expected for each model:
+
+| Base model | resolution|
+|----------------|-----|
+| EfficientNetB0 | 224 |
+| EfficientNetB1 | 240 |
+| EfficientNetB2 | 260 |
+| EfficientNetB3 | 300 |
+| EfficientNetB4 | 380 |
+| EfficientNetB5 | 456 |
+| EfficientNetB6 | 528 |
+| EfficientNetB7 | 600 |
+
+When the model is intended for transfer learning, the Keras implementation
+provides a option to remove the top layers:
+```
+model = EfficientNetB0(include_top=False, weights='imagenet')
+```
+This option excludes the final `Dense` layer that turns 1280 features on the penultimate
+layer into prediction of the 1000 ImageNet classes. Replacing the top layer with custom
+layers allows using EfficientNet as a feature extractor in a transfer learning workflow.
+
+Another argument in the model constructor worth noticing is `drop_connect_rate` which controls
+the dropout rate responsible for [stochastic depth](https://arxiv.org/abs/1603.09382).
+This parameter serves as a toggle for extra regularization in finetuning, but does not
+affect loaded weights. For example, when stronger regularization is desired, try:
+```
+model = EfficientNetB0(weights='imagenet', drop_connect_rate=0.4)
+```
+The default value is 0.2.
+
+## Example: EfficientNetB0 for CIFAR-100.
```

Comment:
Are you sure you still want to use CIFAR-100, as opposed to a dataset with larger natural images, where ImageNet pretraining will be more relevant?

Reply:
I also tried stanford dogs and food 101. They are also good examples. The reasons I am still using CIFAR-100 are: 

1. CIFAR-100 has a transfer learning result that the original EfficientNet paper claimed. So people interested in reproducing the paper may find this example.
2. It is kind of cool that transfer learning allows us to use a high-resolution model on low-resolution data. The original paper even used B7 with 600x600 resolution on CIFAR-100.
3. CIFAR-100 example shows clear improvement on fine-tuning entire model; but this step is not as visible for other datasets like stanford dogs.

However I am completely OK with changing it to stanford dogs for example, and the change is easy.  What do you think?

Reply:
I think the bump in accuracy when fine-tuning the model is due to the fact that the pretrained ImageNet features as not as effective on CIFAR-100 -- so the model trained via feature extraction only is already close to optimal and fine-tuning does not add a lot. Especially for a dogs dataset, since ImageNet contains lots of dogs.

I would recommend using a different dataset, because our code examples should stay as close as possible to the real-world workflows our users are facing. "I want to reproduce the CIFAR-100 results of this specific paper" is not a common use case, but "I have this small dataset of natural images that I want to use to train a classifier" is a common use case :)

Reply:
Got it.  I will change to dogs then. And I think it is actually a good point to mention that low gain in fine-tuning step means most effective transfer learning. 

BTW, since dogs are images of different sizes, would you recommend center crop, random crop or resizing in pipeline (before batching the data)? @fchollet

Reply:
You should resize to a shared size in the pipeline before batching.

You could do this:

````python
  size = (200, 200)
  ds = ds.map(lambda img: tf.image.resize(img, size))
 ```

If the aspect ratio of the images matter, you could do this:

```python
  size = (200, 200)
  ds = ds.map(lambda img: keras.preprocessing.image.smart_resize(img, size))
```

(it will use a combination of cropping and resizing to converse aspect ratio)

`CenterCrop` is something we could optionally do in the model itself, in addition, if the images in the dataset are not well cropped to begin with (too much background). But if the images already look good, this is not necessary.

Reply:
I would really love to try smart_resize, but looks like it is not working with ds that is loaded directly from tfds?

Reply:
That's weird, what error do you get?

Reply:
`OperatorNotAllowedInGraphError: using a `tf.Tensor` as a Python `bool` is not allowed in Graph execution. Use Eager execution or decorate this function with @tf.function.`

Here is a minimal code reproducing the error. It is not really related to tfds actually.
https://colab.research.google.com/drive/1_awoHwxurYy0kM2UNQaUvHuOaMHInRxE?usp=sharing

Reply:
I think I know how to fix this. Will make a PR there.